icicle/include/icicle/fields/field.h

```diff
@@ -802,6 +446,57 @@ class Field
     return r;
   }
 
+  static constexpr HOST_INLINE Field barret_mult(const Field& xs, const Field& ys)
+  {
+    Wide xy = mul_wide(xs, ys); // full mult
+    return reduce(xy);          // reduce mod p
+  }
+
+#endif
+
+  static constexpr HOST_DEVICE_INLINE Field mont_sub_modulus(const Field& xs, bool get_higher = false)
+  {
+    Field r = xs;
+    Field p = Field{get_modulus<1>()};
+    if (p.limbs_storage.limbs[TLC - 1] > r.limbs_storage.limbs[TLC - 1]) return r;
+    ff_storage r_reduced = {};
+    uint64_t carry = 0;
+    carry = sub_limbs<TLC, true>(r.limbs_storage, get_modulus<1>(), r_reduced);
+    if (carry == 0) r = Field{r_reduced};
+    return r;
+  }
+
+  static constexpr HOST_DEVICE_INLINE Field mont_mult(const Field& xs, const Field& ys)
+  {
+    Field r = {};
+    base_math::template multiply_mont<TLC>(
+      xs.limbs_storage, ys.limbs_storage, get_mont_inv_modulus(), get_modulus<1>(), r.limbs_storage);
+    return mont_sub_modulus(r);
+  }
+
+  static constexpr HOST_DEVICE_INLINE Field mont_reduce(const Wide& t)
+  {
+#ifdef __CUDA_ARCH__
```

Comment:
here too

Reply:
waiting for @Koren-Brand fix to the sos function